src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -544,6 +544,14 @@ void QgsOgrProvider::setTransaction( QgsTransaction *transaction )
   QgsDebugMsgLevel( QStringLiteral( "set transaction %1" ).arg( transaction != nullptr ), 2 );
   // static_cast since layers cannot be added to a transaction of a non-matching provider
   mTransaction = static_cast<QgsOgrTransaction *>( transaction );
+  connect( mTransaction, &QgsTransaction::afterRollback, this, [ this ]( )
+  {
+    mFieldsRequireReload = true;
+  } );
+  connect( mTransaction, &QgsTransaction::afterRollbackToSavepoint, this, [ this ]( const QString & )
+  {
+    mFieldsRequireReload = true;
```

Comment:
why not calling loadFields() directly? So far, that's what we were doing, they were no lazy loading of fields and fields were always up to date, so I'm concern about other part of the code that access directly mAttributeFields

Reply:
We can do that, I thought it was more efficient this way: if we are undoing a chain of edits that do multiple rollbacks to savepoints (even  **without** accessing the fields between the individual rollbacks) we would cause multiple calls to `loadFields` without any good reason.

Since this flag `mFieldsRequireReload` is `true` by default and it is only flipped when a rollback is executed I don't think there is any risk involved but if you insist I can remove the flag and just call `loadFields` as you suggested: there shouldn't be any practical difference.

Reply:
> we are undoing a chain of edits that do multiple rollbacks to savepoints (even without accessing the fields between the individual rollbacks) we would cause multiple calls to loadFields without any good reason.

Fair point.

But I'm counting approximatively 20 access to mAttributeField in QgsOgrProvider where loadFields is not called before. If any of that code is called after rollback there may be issues no?

May I suggest to replace direct call of mAtributeFields with fields() while keeping the lazy loading of fields?

Reply:
> > we are undoing a chain of edits that do multiple rollbacks to savepoints (even without accessing the fields between the individual rollbacks) we would cause multiple calls to loadFields without any good reason.
> 
> Fair point.
> 
> But I'm counting approximatively 20 access to mAttributeField in QgsOgrProvider where loadFields is not called before. If any of that code is called after rollback there may be issues no?
> 
> May I suggest to replace direct call of mAtributeFields with fields() while keeping the lazy loading of fields?

that makes sense ... I'll have a look.

Reply:
@troopa81 please have a look to the recent commits: I have checked all direct accesses to mAttributeField and called fields() when required. 

I have also replaced the rollback lambdas with standard signal-slot connections, it looks clearer to me.